{
  "cve_id": "CVE-2021-29555",
  "cve_desc": "TensorFlow is an end-to-end open source platform for machine learning. An attacker can cause a denial of service via a FPE runtime error in `tf.raw_ops.FusedBatchNorm`. This is because the implementation(https://github.com/tensorflow/tensorflow/blob/828f346274841fa7505f7020e88ca36c22e557ab/tensorflow/core/kernels/fused_batch_norm_op.cc#L295-L297) performs a division based on the last dimension of the `x` tensor. Since this is controlled by the user, an attacker can trigger a denial of service. The fix will be included in TensorFlow 2.5.0. We will also cherrypick this commit on TensorFlow 2.4.2, TensorFlow 2.3.3, TensorFlow 2.2.3 and TensorFlow 2.1.4, as these are also affected and still in supported range.",
  "repo": "tensorflow/tensorflow",
  "patch_hash": "1a2a87229d1d61e23a39373777c056161eb4084d",
  "patch_info": {
    "commit_hash": "1a2a87229d1d61e23a39373777c056161eb4084d",
    "repo": "tensorflow/tensorflow",
    "commit_url": "https://github.com/tensorflow/tensorflow/commit/1a2a87229d1d61e23a39373777c056161eb4084d",
    "files": [
      "tensorflow/core/kernels/fused_batch_norm_op.cc"
    ],
    "message": "Fix FPE issue with `tf.raw_ops.FusedBatchNorm`.\n\nPiperOrigin-RevId: 370948185\nChange-Id: If0c8e0320062ed6363e94ff5fe38e6a301f69ac2",
    "before_after_code_files": [
      "tensorflow/core/kernels/fused_batch_norm_op.cc||tensorflow/core/kernels/fused_batch_norm_op.cc"
    ]
  },
  "patch_diff": {
    "tensorflow/core/kernels/fused_batch_norm_op.cc||tensorflow/core/kernels/fused_batch_norm_op.cc": [
      "File: tensorflow/core/kernels/fused_batch_norm_op.cc -> tensorflow/core/kernels/fused_batch_norm_op.cc",
      "--- Hunk 1 ---",
      "[Context before]",
      "293:     const CPUDevice& d = context->eigen_device<CPUDevice>();",
      "295:     const int depth = x.dimension(3);",
      "296:     const int size = x.size();",
      "297:     const int rest_size = size / depth;",
      "298:     Eigen::DSizes<Eigen::Index, 2> rest_by_depth(rest_size, depth);",
      "",
      "[Removed Lines]",
      "[None]",
      "",
      "[Added Lines]",
      "296:     OP_REQUIRES(",
      "297:         context, depth != 0,",
      "298:         errors::Internal(\"The 4th element in the input shape cannot be 0.\"));",
      "",
      "---------------"
    ]
  },
  "candidates": [
    {
      "candidate_hash": "9b3608d2e8aaba75c27f732c70868d6a3ab904c5",
      "candidate_info": {
        "commit_hash": "9b3608d2e8aaba75c27f732c70868d6a3ab904c5",
        "repo": "tensorflow/tensorflow",
        "commit_url": "https://github.com/tensorflow/tensorflow/commit/9b3608d2e8aaba75c27f732c70868d6a3ab904c5",
        "files": [
          "tensorflow/core/kernels/fused_batch_norm_op.cc"
        ],
        "message": "Fix FPE issue with `tf.raw_ops.FusedBatchNorm`.\n\nPiperOrigin-RevId: 370948185\nChange-Id: If0c8e0320062ed6363e94ff5fe38e6a301f69ac2",
        "before_after_code_files": [
          "tensorflow/core/kernels/fused_batch_norm_op.cc||tensorflow/core/kernels/fused_batch_norm_op.cc"
        ]
      },
      "candidate_patch_features": {
        "candidate_earlier_than_patch": 0,
        "diff_branch_same_aad": 1,
        "olp_code_files": {
          "patch": [
            "tensorflow/core/kernels/fused_batch_norm_op.cc||tensorflow/core/kernels/fused_batch_norm_op.cc"
          ],
          "candidate": [
            "tensorflow/core/kernels/fused_batch_norm_op.cc||tensorflow/core/kernels/fused_batch_norm_op.cc"
          ]
        }
      },
      "candidate_diff": {
        "tensorflow/core/kernels/fused_batch_norm_op.cc||tensorflow/core/kernels/fused_batch_norm_op.cc": [
          "File: tensorflow/core/kernels/fused_batch_norm_op.cc -> tensorflow/core/kernels/fused_batch_norm_op.cc",
          "--- Hunk 1 ---",
          "[Context before]",
          "293:     const CPUDevice& d = context->eigen_device<CPUDevice>();",
          "295:     const int depth = x.dimension(3);",
          "296:     const int size = x.size();",
          "297:     const int rest_size = size / depth;",
          "298:     Eigen::DSizes<Eigen::Index, 2> rest_by_depth(rest_size, depth);",
          "",
          "[Removed Lines]",
          "[None]",
          "",
          "[Added Lines]",
          "296:     OP_REQUIRES(",
          "297:         context, depth != 0,",
          "298:         errors::Internal(\"The 4th element in the input shape cannot be 0.\"));",
          "",
          "---------------"
        ]
      }
    },
    {
      "candidate_hash": "97c61cf1a50735605cf153b26fb5c24ae88732b9",
      "candidate_info": {
        "commit_hash": "97c61cf1a50735605cf153b26fb5c24ae88732b9",
        "repo": "tensorflow/tensorflow",
        "commit_url": "https://github.com/tensorflow/tensorflow/commit/97c61cf1a50735605cf153b26fb5c24ae88732b9",
        "files": [
          "tensorflow/core/kernels/fused_batch_norm_op.cc"
        ],
        "message": "Fix FPE issue with `tf.raw_ops.FusedBatchNorm`.\n\nPiperOrigin-RevId: 370948185\nChange-Id: If0c8e0320062ed6363e94ff5fe38e6a301f69ac2",
        "before_after_code_files": [
          "tensorflow/core/kernels/fused_batch_norm_op.cc||tensorflow/core/kernels/fused_batch_norm_op.cc"
        ]
      },
      "candidate_patch_features": {
        "candidate_earlier_than_patch": 0,
        "diff_branch_same_aad": 1,
        "olp_code_files": {
          "patch": [
            "tensorflow/core/kernels/fused_batch_norm_op.cc||tensorflow/core/kernels/fused_batch_norm_op.cc"
          ],
          "candidate": [
            "tensorflow/core/kernels/fused_batch_norm_op.cc||tensorflow/core/kernels/fused_batch_norm_op.cc"
          ]
        }
      },
      "candidate_diff": {
        "tensorflow/core/kernels/fused_batch_norm_op.cc||tensorflow/core/kernels/fused_batch_norm_op.cc": [
          "File: tensorflow/core/kernels/fused_batch_norm_op.cc -> tensorflow/core/kernels/fused_batch_norm_op.cc",
          "--- Hunk 1 ---",
          "[Context before]",
          "293:     const CPUDevice& d = context->eigen_device<CPUDevice>();",
          "295:     const int depth = x.dimension(3);",
          "296:     const int size = x.size();",
          "297:     const int rest_size = size / depth;",
          "298:     Eigen::DSizes<Eigen::Index, 2> rest_by_depth(rest_size, depth);",
          "",
          "[Removed Lines]",
          "[None]",
          "",
          "[Added Lines]",
          "296:     OP_REQUIRES(",
          "297:         context, depth != 0,",
          "298:         errors::Internal(\"The 4th element in the input shape cannot be 0.\"));",
          "",
          "---------------"
        ]
      }
    },
    {
      "candidate_hash": "773ff26570689293210ce910a8c3d8d8c46d4e9a",
      "candidate_info": {
        "commit_hash": "773ff26570689293210ce910a8c3d8d8c46d4e9a",
        "repo": "tensorflow/tensorflow",
        "commit_url": "https://github.com/tensorflow/tensorflow/commit/773ff26570689293210ce910a8c3d8d8c46d4e9a",
        "files": [
          "tensorflow/core/kernels/fused_batch_norm_op.cc"
        ],
        "message": "Fix FPE issue with `tf.raw_ops.FusedBatchNorm`.\n\nPiperOrigin-RevId: 370948185\nChange-Id: If0c8e0320062ed6363e94ff5fe38e6a301f69ac2",
        "before_after_code_files": [
          "tensorflow/core/kernels/fused_batch_norm_op.cc||tensorflow/core/kernels/fused_batch_norm_op.cc"
        ]
      },
      "candidate_patch_features": {
        "candidate_earlier_than_patch": 0,
        "diff_branch_same_aad": 1,
        "olp_code_files": {
          "patch": [
            "tensorflow/core/kernels/fused_batch_norm_op.cc||tensorflow/core/kernels/fused_batch_norm_op.cc"
          ],
          "candidate": [
            "tensorflow/core/kernels/fused_batch_norm_op.cc||tensorflow/core/kernels/fused_batch_norm_op.cc"
          ]
        }
      },
      "candidate_diff": {
        "tensorflow/core/kernels/fused_batch_norm_op.cc||tensorflow/core/kernels/fused_batch_norm_op.cc": [
          "File: tensorflow/core/kernels/fused_batch_norm_op.cc -> tensorflow/core/kernels/fused_batch_norm_op.cc",
          "--- Hunk 1 ---",
          "[Context before]",
          "293:     const CPUDevice& d = context->eigen_device<CPUDevice>();",
          "295:     const int depth = x.dimension(3);",
          "296:     const int size = x.size();",
          "297:     const int rest_size = size / depth;",
          "298:     Eigen::DSizes<Eigen::Index, 2> rest_by_depth(rest_size, depth);",
          "",
          "[Removed Lines]",
          "[None]",
          "",
          "[Added Lines]",
          "296:     OP_REQUIRES(",
          "297:         context, depth != 0,",
          "298:         errors::Internal(\"The 4th element in the input shape cannot be 0.\"));",
          "",
          "---------------"
        ]
      }
    },
    {
      "candidate_hash": "5ae65f75c50188468667f2e3a49b8ff2c6ab30ec",
      "candidate_info": {
        "commit_hash": "5ae65f75c50188468667f2e3a49b8ff2c6ab30ec",
        "repo": "tensorflow/tensorflow",
        "commit_url": "https://github.com/tensorflow/tensorflow/commit/5ae65f75c50188468667f2e3a49b8ff2c6ab30ec",
        "files": [
          "tensorflow/core/kernels/fused_batch_norm_op.cc"
        ],
        "message": "Fix FPE issue with `tf.raw_ops.FusedBatchNorm`.\n\nPiperOrigin-RevId: 370948185\nChange-Id: If0c8e0320062ed6363e94ff5fe38e6a301f69ac2",
        "before_after_code_files": [
          "tensorflow/core/kernels/fused_batch_norm_op.cc||tensorflow/core/kernels/fused_batch_norm_op.cc"
        ]
      },
      "candidate_patch_features": {
        "candidate_earlier_than_patch": 0,
        "diff_branch_same_aad": 1,
        "olp_code_files": {
          "patch": [
            "tensorflow/core/kernels/fused_batch_norm_op.cc||tensorflow/core/kernels/fused_batch_norm_op.cc"
          ],
          "candidate": [
            "tensorflow/core/kernels/fused_batch_norm_op.cc||tensorflow/core/kernels/fused_batch_norm_op.cc"
          ]
        }
      },
      "candidate_diff": {
        "tensorflow/core/kernels/fused_batch_norm_op.cc||tensorflow/core/kernels/fused_batch_norm_op.cc": [
          "File: tensorflow/core/kernels/fused_batch_norm_op.cc -> tensorflow/core/kernels/fused_batch_norm_op.cc",
          "--- Hunk 1 ---",
          "[Context before]",
          "293:     const CPUDevice& d = context->eigen_device<CPUDevice>();",
          "295:     const int depth = x.dimension(3);",
          "296:     const int size = x.size();",
          "297:     const int rest_size = size / depth;",
          "298:     Eigen::DSizes<Eigen::Index, 2> rest_by_depth(rest_size, depth);",
          "",
          "[Removed Lines]",
          "[None]",
          "",
          "[Added Lines]",
          "296:     OP_REQUIRES(",
          "297:         context, depth != 0,",
          "298:         errors::Internal(\"The 4th element in the input shape cannot be 0.\"));",
          "",
          "---------------"
        ]
      }
    },
    {
      "candidate_hash": "9cf9f990ce00f5808406f70cf6cc0d3a6b97d45e",
      "candidate_info": {
        "commit_hash": "9cf9f990ce00f5808406f70cf6cc0d3a6b97d45e",
        "repo": "tensorflow/tensorflow",
        "commit_url": "https://github.com/tensorflow/tensorflow/commit/9cf9f990ce00f5808406f70cf6cc0d3a6b97d45e",
        "files": [
          "tensorflow/core/kernels/fused_batch_norm_op.cc"
        ],
        "message": "Fix FPE issue with `tf.raw_ops.FusedBatchNorm`.\n\nPiperOrigin-RevId: 370948185\nChange-Id: If0c8e0320062ed6363e94ff5fe38e6a301f69ac2",
        "before_after_code_files": [
          "tensorflow/core/kernels/fused_batch_norm_op.cc||tensorflow/core/kernels/fused_batch_norm_op.cc"
        ]
      },
      "candidate_patch_features": {
        "candidate_earlier_than_patch": 0,
        "diff_branch_same_aad": 1,
        "olp_code_files": {
          "patch": [
            "tensorflow/core/kernels/fused_batch_norm_op.cc||tensorflow/core/kernels/fused_batch_norm_op.cc"
          ],
          "candidate": [
            "tensorflow/core/kernels/fused_batch_norm_op.cc||tensorflow/core/kernels/fused_batch_norm_op.cc"
          ]
        }
      },
      "candidate_diff": {
        "tensorflow/core/kernels/fused_batch_norm_op.cc||tensorflow/core/kernels/fused_batch_norm_op.cc": [
          "File: tensorflow/core/kernels/fused_batch_norm_op.cc -> tensorflow/core/kernels/fused_batch_norm_op.cc",
          "--- Hunk 1 ---",
          "[Context before]",
          "301:     const CPUDevice& d = context->eigen_device<CPUDevice>();",
          "303:     const int depth = x.dimension(3);",
          "304:     const int size = x.size();",
          "305:     const int rest_size = size / depth;",
          "306:     Eigen::DSizes<Eigen::Index, 2> rest_by_depth(rest_size, depth);",
          "",
          "[Removed Lines]",
          "[None]",
          "",
          "[Added Lines]",
          "304:     OP_REQUIRES(",
          "305:         context, depth != 0,",
          "306:         errors::Internal(\"The 4th element in the input shape cannot be 0.\"));",
          "",
          "---------------"
        ]
      }
    }
  ]
}